consensus/state_sync_handler.go

```diff
@@ -5,59 +5,172 @@ import (
-
 	switch stateSyncMessageAny.MessageName() {
 	case messaging.StateSyncMessageContentType:
 		msg, err := codec.GetCodec().FromAny(stateSyncMessageAny)
 		if err != nil {
 			return err
 		}
-
 		stateSyncMessage, ok := msg.(*typesCons.StateSyncMessage)
 		if !ok {
 			return fmt.Errorf("failed to cast message to StateSyncMessage")
 		}
-
 		return m.handleStateSyncMessage(stateSyncMessage)
+
 	default:
 		return typesCons.ErrUnknownStateSyncMessageType(stateSyncMessageAny.MessageName())
 	}
 }
 
 func (m *consensusModule) handleStateSyncMessage(stateSyncMessage *typesCons.StateSyncMessage) error {
 	switch stateSyncMessage.Message.(type) {
+
 	case *typesCons.StateSyncMessage_MetadataReq:
 		m.logger.Info().Str("proto_type", "MetadataRequest").Msg("Handling StateSyncMessage MetadataReq")
-		if !m.serverModeEnabled {
+		if !m.consCfg.ServerModeEnabled {
 			m.logger.Warn().Msg("Node's server module is not enabled")
 			return nil
 		}
-		return m.stateSync.HandleStateSyncMetadataRequest(stateSyncMessage.GetMetadataReq())
-	case *typesCons.StateSyncMessage_MetadataRes:
-		m.logger.Info().Str("proto_type", "MetadataResponse").Msg("Handling StateSyncMessage MetadataRes")
-		m.metadataReceived <- stateSyncMessage.GetMetadataRes()
+		go m.stateSync.HandleStateSyncMetadataRequest(stateSyncMessage.GetMetadataReq())
 		return nil
+
 	case *typesCons.StateSyncMessage_GetBlockReq:
 		m.logger.Info().Str("proto_type", "GetBlockRequest").Msg("Handling StateSyncMessage GetBlockRequest")
-		if !m.serverModeEnabled {
+		if !m.consCfg.ServerModeEnabled {
 			m.logger.Warn().Msg("Node's server module is not enabled")
 			return nil
 		}
-		return m.stateSync.HandleGetBlockRequest(stateSyncMessage.GetGetBlockReq())
+		go m.stateSync.HandleGetBlockRequest(stateSyncMessage.GetGetBlockReq())
+		return nil
+
+	case *typesCons.StateSyncMessage_MetadataRes:
+		m.logger.Info().Str("proto_type", "MetadataResponse").Msg("Handling StateSyncMessage MetadataRes")
+		go m.stateSync.HandleStateSyncMetadataResponse(stateSyncMessage.GetMetadataRes())
+		return nil
+
+	// NB: Note that this is the only case that calls a function in the consensus module (not the state sync submodule) since
+	// consensus is the one responsible for calling business logic to apply and commit the blocks. State sync listens for events
+	// that are a result of it.
 	case *typesCons.StateSyncMessage_GetBlockRes:
 		m.logger.Info().Str("proto_type", "GetBlockResponse").Msg("Handling StateSyncMessage GetBlockResponse")
-		m.blocksReceived <- stateSyncMessage.GetGetBlockRes()
+		go m.tryToApplyRequestedBlock(stateSyncMessage.GetGetBlockRes())
 		return nil
+
 	default:
 		return fmt.Errorf("unspecified state sync message type")
 	}
 }
+
+// tryToApplyRequestedBlock tries to commit the requested Block received from a peer.
+// Intended to be called via a background goroutine.
+// CLEANUP: Investigate whether this should be part of `Consensus` or part of `StateSync`
```

Comment:
IMO, It should be a block application module, that both depend on. It will be applying blocks when either of them is active.

Reply:
The core business logic for applying blocks is in the `Utility` module, but this is an interesting idea. 

I like this idea though. I can see us creating a ticket to have a single ticket that refactors things to have a Block Application Module after we get the existing implementation working on main.